### ISSUE TITLE
Bug in the projection layer

### ISSUE THREAD

**Issue by kylin0421** [filipbasara0/matryoshka-representation-learning#1]
Hi, there seems to be a bug in Matryoshka Projection layer which makes it equivalent to a single linear layer:

class MatryoshkaProjector(nn.Module):
    def __init__(self, nesting_dims, out_dims, **kwargs):
        super(MatryoshkaProjector, self).__init__()
        self.nesting_dims = nesting_dims
        self.proj_hidden = nn.Linear(nesting_dims[-1], nesting_dims[-1], **kwargs)
        self.relu = nn.ReLU()
        self.proj_linear = nn.Linear(nesting_dims[-1], out_dims, **kwargs)

    def _apply_linear_layer(self, x, layer, nesting_dim):
        logits = torch.matmul(x[:, :nesting_dim], layer.weight[:, :nesting_dim].t())
        if layer.bias != None:
            logits += layer.bias
        return logits

    def forward(self, x):
        nesting_logits = []
        for nesting_dim in self.nesting_dims:
            logits = self._apply_linear_layer(x, self.proj_hidden, nesting_dim)
            logits = self.relu(logits)
            logits = self._apply_linear_layer(x, self.proj_linear, nesting_dim)
            nesting_logits.append(logits)
        return nesting_logits

In the forward(), logits is set to self._apply_linear_layer(x, self.proj_linear, nesting_dim) finally, which logically skips the hidden layer and relu.

**Comment by kylin0421** [comment]
Ok, it seems adding a hidden layer will not increase the performance.
For each results set, the upper one is MP layer with mlp, the lower one is MP layer with one single linear:


Accuracy (8):  0.5881
Accuracy (8):  0.6363
Accuracy (16):  0.6436
Accuracy (16):  0.6899
Accuracy (32):  0.6848
Accuracy (32):  0.728
Accuracy (64):  0.6931
Accuracy (64):  0.7372
Accuracy (128):  0.7053
Accuracy (128):  0.7473
Accuracy (256):  0.7162
Accuracy (256):  0.7603
Accuracy (512):  0.7255
Accuracy (512):  0.7701
